connection_e2e_test.go

```diff
@@ -314,3 +316,145 @@ func TestConnection(t *testing.T) {
 		}
 	})
 }
+
+func BenchmarkWithCertificateCache(b *testing.B) {
+	b.ReportAllocs()
+
+	bytes, err := os.ReadFile(os.Getenv("YDB_SSL_ROOT_CERTIFICATES_FILE"))
+	if err != nil {
+		b.Fatal(err)
+	}
+	os.Unsetenv("YDB_SSL_ROOT_CERTIFICATES_FILE")
+
+	ctx := context.TODO()
+	db, err := ydb.Open(
+		ctx,
+		os.Getenv("YDB_CONNECTION_STRING"),
+		ydb.WithCertificatesFromPem(bytes),
+	)
+	if err != nil {
+		b.Fatal(err)
+	}
+	defer func() {
+		// cleanup connection
+		if e := db.Close(ctx); e != nil {
+			b.Fatalf("close failed: %+v", e)
+		}
+	}()
+
+	tcs := []struct {
+		name    string
+		enabled bool
+	}{
+		{"no cache", false},
+		{"cached", true},
+	}
+	for _, tc := range tcs {
+		b.Run(tc.name, func(b *testing.B) {
+			cached.CertificateCache().Enable(tc.enabled)
+			for i := 0; i < b.N; i++ {
+				// child conns are closed on db.Close()
+				_, err := db.With(
+					ctx,
+					ydb.WithSessionPoolSizeLimit(100),
+				)
+				if err != nil {
+					b.Fatal(err)
+				}
+			}
+		})
+	}
+}
+
+func BenchmarkWithFileCache(b *testing.B) {
+	b.ReportAllocs()
+
+	ctx := context.TODO()
+	db, err := ydb.Open(
+		ctx,
+		os.Getenv("YDB_CONNECTION_STRING"),
+	)
+	if err != nil {
+		b.Fatal(err)
+	}
+	defer func() {
+		// cleanup connection
+		if e := db.Close(ctx); e != nil {
+			b.Fatalf("close failed: %+v", e)
+		}
+	}()
+
+	tcs := []struct {
+		name        string
+		certEnabled bool
+		fileEnabled bool
+	}{
+		{"no cache", false, false},
+		{"cert cache", true, false},
+		{"both caches", true, true},
+	}
+	for _, tc := range tcs {
+		b.Run(tc.name, func(b *testing.B) {
+			cached.FileCache().Enable(tc.fileEnabled)
+			cached.FileCache().Reset()
+
+			cached.CertificateCache().Enable(tc.certEnabled)
+			cached.CertificateCache().Reset()
+
+			b.ResetTimer()
+
+			for i := 0; i < b.N; i++ {
+				// child conns are closed on db.Close()
+				_, err := db.With(
+					ctx,
+					ydb.WithSessionPoolSizeLimit(100),
+				)
+				if err != nil {
+					b.Fatal(err)
+				}
+			}
+		})
+	}
+}
+
+func TestWithCacheConcurrent(t *testing.T) {
+	const nRoutines = 10
+
+	ctx := context.TODO()
+	db, err := ydb.Open(
+		ctx,
+		os.Getenv("YDB_CONNECTION_STRING"),
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		// cleanup connection
+		if e := db.Close(ctx); e != nil {
+			t.Fatalf("close failed: %+v", e)
+		}
+	}()
+
+	wg := &sync.WaitGroup{}
+
+	for i := 0; i < nRoutines; i++ {
+		wg.Add(1)
+		go func() {
+			// child conns are closed on db.Close()
+			_, err := db.With(
+				ctx,
+				ydb.WithSessionPoolSizeLimit(100),
+			)
+			if err != nil {
+				t.Error(err)
+			}
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+
+	want, got := nRoutines, cached.FileCache().Hits()
+	if want != got {
```

Comment:
This test is not good enough: it [fails](https://github.com/ydb-platform/ydb-go-sdk/actions/runs/3255017774/jobs/5343923180) on reconnect because certs are loaded again. Maybe should change to got >= want because we really only care about cache misses on child connections.